### DISCUSSION TITLE
Support for ROS 2 Type Description Distribution for the Micro-XRCE-DDS-Agent & Client?

### DISCUSSION THREAD

**Discussion by bkueng** [eProsima/Micro-XRCE-DDS#160] [Ideas]
I was wondering if there are any plans to implement the [ROS 2 type information](https://github.com/ros2/ros2/issues/1159) into the Micro XRCE client and agent, or whether it's preferred to keep the implementation agnostic to anything ROS 2-specific?
In particular I'm interested in a way to check for message format compatibility over the bridge.

**Comment by pablogs9** [comment]
Hello, @bkueng as far as I understand ROS 2 type information at the middleware level only implies adding a sort of hash at the end of the mangled type name.

In XRCE-DDS (as in DDS), the user is in charge of determining its topic and type names. In general, XRCE-DDS will be as ROS 2 agnostic as possible.

> In particular I'm interested in a way to check for message format compatibility over the bridge.

The DDS standard assumes that two endpoints that use the same topic and type name are coherent in their type.
The X-Types standard address this issue by moving type information through the wire, but unfortunately, XRCE-DDS does not work properly with X-Types.

**Reply by bkueng** [reply]
I'm not very familiar with it, but my understanding is that adding the hash to the type name got changed to using USER_DATA instead (https://github.com/ros2/rmw_fastrtps/pull/671).

> The X-Types standard address this issue by moving type information through the wire, but unfortunately, XRCE-DDS does not work properly with X-Types.

Can you elaborate why X-Types do not work properly here?

**Reply by pablogs9** [reply]
Simply because they are not implemented in XRCE-DDS Client and Agent.
